lib/pbench/server/indexer.py

```diff
@@ -106,31 +106,33 @@ def _sleep_w_backoff(backoff):
     _sleep(_calc_backoff_sleep(backoff))
 
 
-def _get_es_hosts(config, logger):
+def _get_es_hosts(config):
     """
     Return list of dicts (a single dict for now) - that's what ES is expecting.
     """
     try:
         uri = config.get("Indexing", "uri")
     except (configparser.NoSectionError, configparser.NoOptionError):
         raise BadConfig("Indexing URI missing")
+
     url = urlparse(uri)
     if not url.hostname:
         raise BadConfig("Indexing URI must contain a host name")
-    if not url.port:
-        raise BadConfig("Indexing URI must contain a port number")
+
     timeoutobj = Timeout(total=1200, connect=10, read=_read_timeout)
-    return [
-        dict(host=url.hostname, port=url.port, timeout=timeoutobj),
-    ]
+    rv = {"host": url.hostname, "timeout": timeoutobj}
+    if url.port:
+        rv["port"] = url.port
+
+    return [rv]
```

Comment:
I would just drop the `urlparse` entirely and use exactly what we get from `[Indexing] uri` as-is instead of breaking it up into fields. I suppose the downside is that I can't see any way to pass a `Timeout` object through to the connection layer without doing the full `dict` breakdown ahead of time.

But if we're going to do the `dict`, I think you need to do the full logic because it doesn't look like `Elasticsearch` does anything to "break down" a `dict`'s `"host"` field.

E.g., if you pass just `["<url>"]` to `Elasticsearch`, it ends up doing this:

```python
            parsed_url = urlparse(host)
            h = {"host": parsed_url.hostname}

            if parsed_url.port:
                h["port"] = parsed_url.port

            if parsed_url.scheme == "https":
                h["port"] = parsed_url.port or 443
                h["use_ssl"] = True

            if parsed_url.username or parsed_url.password:
                h["http_auth"] = "%s:%s" % (
                    unquote(parsed_url.username),
                    unquote(parsed_url.password),
                )

            if parsed_url.path and parsed_url.path != "/":
                h["url_prefix"] = parsed_url.path
```

Reply:
Yeah, I meant to comment on this and ask you this very question, but I got distracted, and then I got distracted from my distraction, and now you're distracting me from that....

I _thought_ there was an option for passing an unparsed URL; there's no way to combine that with a `Timeout` object?

Reply:
So far as I can see, it's "all or nothing". You can pass a string, or you can pass a fully broken-down `dict`, and nothing in between. The constructor just looks for a string type and breaks it down for you (as I pasted above): but if it's already a `dict` it just passes it on.